### ISSUE TITLE
Make it easier to pass scopes in to scope/3 callbacks

### ISSUE THREAD

**Issue by schrockwell** [schrockwell/bodyguard#15]
Right now it's technically possible through the optional `opts` list, but it would be nice if there was a convention for it so it was more obvious.

**Comment by schrockwell** [comment]
I'm thinking about changing the function definition of `scope(user, action, opts)` to `scope(user, scope, opts)` and just assuming that the scope is independent of what action is being authorized. 

The new the `scope` parameter is a queryable to be passed in by the caller, similar to how Pundit's `Scope` class expects a scope to be passed into it.

I guess we'd also have to think about how it was called, because the controller helper `scope(conn, Post, Post)` seems pretty silly (one being the schema and one being the provided queryable). Maybe there could be a `scope(conn, Post)` which short-cuts to `scope(conn, Post, Post)`.

**Comment by schrockwell** [comment]
Alternatively, we just change the callback to the even shorter `scope(user, opts)` and drop the scope requirement completely. If a user wants to pass in an explicit "parent" scope, they can do so in the `opts` keyword list. So the controller case would look like `scope(conn, Post, query: Post)` and then the `scope/2` callback in the policy module would look at `opts[:query]`.

Or, going the other direction, change it to `scope(user, scope)` and omit the `opts` entirely.

Thoughts on this or the above comment, @bencates?

**Comment by abitdodgy** [comment]
I think being able to pass in the scope is key. At the moment, I'm doing this:

```elixir
  def index(conn, _params, _current_user, current_org) do
    invites = scope(conn, Invite, scope: assoc(current_org, :invites)) |> Repo.all
    render conn, :index, invites: invites
  end

  # invite_policy.ex
  def scope(user, _action, opts) do
    scope = Keyword.fetch!(opts, :scope)
    Ecto.Query.where(scope, status: ^:pending)
  end
```

I would rather do something like this unless there's a better way:

```elixir
  def index(conn, _params, _current_user, current_org) do
    invites = scope(conn, assoc(current_org, :invites)) |> Repo.all
    # ...
  end
```

**Comment by schrockwell** [comment]
@abitdodgy Thanks for the feedback, and I tend to agree. I'd like to keep the `action` argument in the `scope` callback as additional flexibility, since it's easy enough to wildcard match it using pattern matching if you don't need it.

I'm planning to change the `scope/3` callback to have the signature `scope(user, action, scope)` and omit `opts` entirely. The scope will be aware of Ecto queries, so you can do any of the following in a controller:

```elixir
# By schema name
scope(conn, Invite)

# By Ecto queries
scope(conn, assoc(current_org, :invites))
scope(conn, from(i in Invites))

# By some other scope where the schema can't be inferred
scope(conn, some_context, policy: Invite.Policy)
```

And the callbacks will be:

```elixir
def scope(user, _action, scope) do
  # Do something with user and scope
end

def scope(_user, _action, scope), do: scope
```

This WILL be a breaking change for v0.6, but that's permitted in SemVer before hitting v1.0, and I've added a warning when the old behavior is detected (i.e. you try to pass extra `opts` that will now be thrown out).

Any objections?

**Comment by schrockwell** [comment]
Implemented in v0.6.0!
